### DISCUSSION TITLE
Async cache, best practices (python3.10)?

### DISCUSSION THREAD

**Discussion by acidjunk** [fastapi/fastapi#6346] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from aiocache import cached
from fastapi.concurrency import run_in_threadpool

from our.schemas import OrganisationSchema
from our.services import crm

router = APIRouter()

@router.get("/organisations", response_model=list[OrganisationSchema])
@cached(alias="crm")
async def organisations() -> list[dict]:
    return await run_in_threadpool(crm.get_organisations)
```

### Description

We (my dev team At SURF) use an async cache solution in a couple of FastAPI apps. It caches complete endpoints and generally works ok for our use cases. But as we upgraded our code base to python3.10 the cache solution broke and [aiocache](https://github.com/aio-libs/aiocache) seems to be [struggling](https://github.com/aio-libs/aiocache/issues/536) to find a new maintainer.

So for now we created a fork that works with 3.10. A couple of options seem to exists:

1) we decide to start maintaining the aiocache lib
2) switch to another cache solution that support our use cases and has a higher chance of continued maintenance

This Question is mostly aimed at 2: Does someone use async caching on an endpoint level? If so are you satisfied with it? Can you elaborate on some best practices?

### Operating System

Linux

### FastAPI Version

0.75.0

### Python Version

3.10

**Comment by lmeyerov** [comment, marked as answer]
@acidjunk actively looking at this, not for fastapi but some other asyncio python server 

- we're especially interested in LRU + TLRU. TLRU is usually expiry defined by insertion time (imagine a expiry header), but we're also interested in expiring based on time since last read, so we can more aggressively drain anything not accessed recently, vs just when maxsize is hit

- race / reentrancy: our servers are generally multiple share-nothing processes, and each process is an asyncio greenthreaded event loop. where this gets a bit funny is if there is something like `await slow_route('some_key')`, two concurrent requests may ask for `'some_key'`, and you want to make sure only the first is dispatched, and the second knows to wait . Likewise, should be safe for a concurrent `'some_key2'` to proceed

- ...some people are doing multithreaded caches, but gh issues look like maintenance nightmares and scare off contributors. one neat trick has been exposing optional `lock=...` params, and I can imagine some sort of `rw_lock_class=` and `lock=` modes... but again... seems like a can of worms

- some flexibility questions like whether to cache at route level or across routes, and external controls for tuning specific routes

- our world is a bit more interesting b/c we're dealing with both GPU + CPU buffers as part of multi-stage compute, so we might have a nesting where we do N GPU cached objects and 5 * N CPU cached objects

- we do cache a bunch at the client & proxy server level via smart clients & content headers, but internally we do end up wanting a bit more incrementalization, like `big_compute?sort=True&filter=x` where might want to internally cache `sort=True` and then just on-the-fly `filter` (simplified example), and thus cannot use the client/proxy tiers


As far as I can tell, all python caching libraries are dragging their feet on asyncio with basic LRU/TLRU that we'd expect for a web server . Doing fastapi decorators as plugins/wrappers over a common core may be the more sustainable mode as it'd guarantee fastapi use and enable reuse among non-fastapi ones (`cached`, `async_lru`, `aiocache`, ...)

**Comment by acidjunk** [comment]
@lmeyerov Thanks!

Will leave the ticket open for a couple of weeks; we will probably decide to put some own efforts in aiocache for the time being.

**Comment by acidjunk** [comment]
Closing it. Not sure yet on the best path forward.

**Comment by acidjunk** [comment]
We decided to write our own implementation:
https://github.com/workfloworchestrator/nwa-stdlib/pull/11
